### PR TITLE
Exfat next

### PR DESCRIPTION
**Pull request by namjaejeon**

### CODE REVIEW

File: dump/dump.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+		exfat_err("bitmap entry read failed: %d\n", errno);
+		ret = -EIO;
+		goto free_entry;
+	}
+
+	volume_label = exfat_conv_volume_serial(&ed[0]);
+	if (!volume_label)
+		return -EINVAL;
+
+	bitmap_clu = le32_to_cpu(ed[1].bitmap_start_clu);
```
Comment (hclee) [inline dump/dump.c:155]: Shouldn't we check the type of the directory entry?

File: dump/dump.c [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+	if (!bitmap) {
+		exfat_err("bitmap allocation failed\n");
+		goto free_entry;
+	}
+
+	ret = exfat_read(bd->dev_fd, bitmap, bitmap_len, bitmap_clu_off);
+	if (ret < 0) {
+		exfat_err("bitmap entry read failed: %d\n", errno);
+		ret = -EIO;
+		goto free_entry;
```
Comment (hclee) [inline dump/dump.c:185]: We need to free the bitmap.

File: dump/dump.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+	ret = exfat_read(bd->dev_fd, ed, sizeof(struct exfat_dentry)*3,
+			root_clu_off);
+	if (ret < 0) {
+		exfat_err("bitmap entry read failed: %d\n", errno);
+		ret = -EIO;
+		goto free_entry;
+	}
+
+	volume_label = exfat_conv_volume_serial(&ed[0]);
```
Comment (hclee) [inline dump/dump.c:151]: We need to free volume_label somewhere.

File: lib/libexfat.c [hunk 4]
```diff
@@ -604,3 +621,10 @@ int exfat_set_volume_serial(struct exfat_blk_dev *bd,
 
 	return ret;
 }
+
+unsigned int exfat_calc_cluster_offset(struct exfat_blk_dev *bd,
+		unsigned int clu_off, unsigned int clu)
```
Comment (hclee) [inline lib/libexfat.c:626]: I think the names of this function and  arguments are not clear. Shouldn't we reflect the device byte offset in the function name and sector unit in the clu_off argument?

File: dump/dump.c [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+}
+
+static struct option opts[] = {
+	{"version",		no_argument,		NULL,	'V' },
+	{"help",		no_argument,		NULL,	'h' },
+	{"?",			no_argument,		NULL,	'?' },
+	{NULL,			0,			NULL,	 0  }
+};
+
+static int exfat_count_used_clusters(unsigned char *bitmap,
```
[the remark this answers is not in the document]
Comment (namjaejeon) [reply dump/dump.c:55]: Okay.

File: dump/dump.c [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+	memset(ui.dev_name, 0, sizeof(ui.dev_name));
+	snprintf(ui.dev_name, sizeof(ui.dev_name), "%s", argv[1]);
+
+	ret = exfat_get_blk_dev_info(&ui, &bd);
+	if (ret < 0)
+		goto out;
+
+	ret = exfat_show_ondisk_all_info(&bd);
+
```
[the remark this answers is not in the document]
Comment (namjaejeon) [reply dump/dump.c:245]: Okay.

File: dump/dump.c [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+		return -1;
+	}
+
+	/* read main boot sector */
+	ret = exfat_read_sector(bd, (char *)ppbr, BOOT_SEC_IDX);
+	if (ret < 0) {
+		exfat_err("main boot sector read failed\n");
+		ret = -1;
+		goto free_ppbr;
+	}
```
[the remark this answers is not in the document]
Comment (namjaejeon) [reply dump/dump.c:91]: It just show ondisk metadata info, not mount and read/write some file. It should show information even though exfat is corrupted.

File: dump/dump.c (hunk 1 above)
Comment (namjaejeon) [reply dump/dump.c:155]: As I said, I should show ondisk information even though it is corrupted.

File: dump/dump.c (hunk 2 above)
Comment (namjaejeon) [reply dump/dump.c:185]: Right. will update it.

File: lib/libexfat.c (hunk 4 above)
Comment (namjaejeon) [reply lib/libexfat.c:626]: the function name and clu_off argument will rename to exfat_clus_to_blk_dev_off() and clu_off_sectors.

**Comment by namjaejeon** [comment]
@hclee I have updated your review points. Let me know if you have more comment. Thanks!

**Comment by hclee** [comment]
> @hclee I have updated your review points. Let me know if you have more comment. Thanks!

it looks good to me, but shouldn't we free volume_label allocated by exfat_conv_volume_serial in exfat_show_ondisk_all_info?

**Comment by namjaejeon** [comment]
@hclee Done. Please check updated patches. Thanks for your review!

**Pull request merged**